boot/seal.go

```diff
@@ -0,0 +1,293 @@
+	cmdline, err := ComposeCandidateCommandLine(model)
+	if err != nil {
+		return fmt.Errorf("cannot obtain kernel command line: %v", err)
+	}
+	// Get the expected kernel command line of the recovery system we're installing from
+	recoveryCmdline, err := ComposeRecoveryCommandLine(model, modeenv.RecoverySystem)
+	if err != nil {
+		return fmt.Errorf("cannot obtain recovery kernel command line: %v", err)
+	}
+	kernelCmdlines := []string{
+		cmdline,
+		recoveryCmdline,
+	}
+
+	sealKeyParams := secboot.SealKeyParams{
+		ModelParams: []*secboot.SealKeyModelParams{
+			{
+				Model:          model,
+				KernelCmdlines: kernelCmdlines,
+				EFILoadChains:  append(runModeChains, recoverModeChains...),
+			},
+		},
+		KeyFile:                 filepath.Join(InitramfsEncryptionKeyDir, "ubuntu-data.sealed-key"),
+		TPMPolicyUpdateDataFile: filepath.Join(InstallHostFDEDataDir, "policy-update-data"),
+		TPMLockoutAuthFile:      filepath.Join(InstallHostFDEDataDir, "tpm-lockout-auth"),
+	}
+
+	if err := secbootSealKey(key, &sealKeyParams); err != nil {
+		return fmt.Errorf("cannot seal the encryption key: %v", err)
+	}
+
+	return nil
+}
+
+// recoverModeLoadSequences builds the recovery boot file list based on
+// the boot chain retrieved from the bootloader and trusted assets listed
+// in modeenv.
+func recoverModeLoadSequences(rbl bootloader.Bootloader, modeenv *Modeenv) ([][]bootloader.BootFile, error) {
+	tbl, ok := rbl.(bootloader.TrustedAssetsBootloader)
+	if !ok {
+		return nil, fmt.Errorf("bootloader doesn't support trusted assets")
+	}
+
+	kernelPath, err := kernelPathForRecoverySystem(modeenv.RecoverySystem)
+	if err != nil {
+		return nil, err
+	}
+
+	bootChain, err := tbl.RecoveryBootChain(kernelPath)
+	if err != nil {
+		return nil, err
+	}
+
+	seq0 := make([]bootloader.BootFile, 0, len(bootChain))
+	seq1 := make([]bootloader.BootFile, 0, len(bootChain))
+
+	// walk the chain and get cache entries for the trusted assets
+	for i := 0; i < len(bootChain)-1; i++ {
+		bf := bootChain[i]
+		name := filepath.Base(bf.Path)
+
+		if bf.Role != bootloader.RoleRecovery {
+			return nil, fmt.Errorf("internal error: recovery boot chain contains invalid role %v", bf.Role)
+		}
+
+		p0, p1, err := cachedAssetPathnames(rbl.Name(), name, modeenv.CurrentTrustedRecoveryBootAssets)
+		if err != nil {
+			return nil, err
+		}
+		seq0 = append(seq0, bf.WithPath(p0))
+		seq1 = append(seq1, bf.WithPath(p1))
+	}
+
+	// add the recovery kernel
+	bf := bootChain[len(bootChain)-1]
+	seq0 = append(seq0, bf)
+	seq1 = append(seq1, bf)
+
+	if sequenceEqual(seq0, seq1) {
+		return [][]bootloader.BootFile{seq0}, nil
+	}
+
+	return [][]bootloader.BootFile{seq0, seq1}, nil
+}
+
+// kernelPathForRecoverySystem validates the kernel for the specified
+// recovery system and returns the path to the kernel snap.
+func kernelPathForRecoverySystem(recoverySystem string) (string, error) {
+	systemSeed, err := seed.Open(dirs.SnapSeedDir, recoverySystem)
```

Comment:
mocking seed.Open itself is probably the easiest here given we care about just one snap (the kernel)

Reply:
I have a slightly better plan that would also simplify snap-bootstrap a little bit...

Reply:
#9283 should help  (I forgot to post this one comment)